File: packages/admin/evolve_admin/mcp_bridge/google_tools.py
"""MCP bridge tools for Google services (Gmail / Drive / Calendar).

**Thin delegation layer.** As of the P1 bot-agent tool work
([internal/spec-google-integration-architecture-2026-06-20.md](../../../internal/spec-google-integration-architecture-2026-06-20.md)
§8 decision 4 — one shared module, no forked logic) the curated tool logic
lives in :mod:`evolve_admin.google_service`. The ``tool_*`` handlers here are
the **evolve-pod MCP bridge** surface (Claude Desktop / evo / admin): each
resolves a bot from the bridge's session/registry — the bridge keeps its
historical ``bot``-argument behavior, an *operator* surface — and delegates to
the shared module.

The other surface, the **bot-facing** ``/api/google/*`` admin route, calls the
same shared module but binds the bot identity server-side to the authenticated
caller (never a request field) — see ``web/google_bot_routes.py``. Repointing
this bridge onto the server-bound identity model is a deliberate fast-follow
(spec §8 decision 4 / DEFERRED list); for P1 the bridge is unchanged in
behavior and shares only the tool *logic*.

Fifteen tools: Gmail ×9, Calendar ×2, Drive ×4. Docs / Sheets / Slides are
deferred to P3.

The bridge runs as the ``evolve`` user, so it has read access to the SA-JSON
secrets and OAuth tokens that ``google_auth.load_credentials`` resolves; bots
never see either — they call tools which load credentials in-process and make
the API calls server-side.

Re-exports below keep ``google_tools.<helper>`` import paths and monkeypatch
targets valid for callers/tests that referenced them before the extract.
"""

from __future__ import annotations

from .. import config, google_auth  # noqa: F401 — re-exported (test monkeypatch targets)
from .. import google_service
from .registry import BotRegistry

# Re-export the shared module's helpers + scope constants so existing import
# paths (``google_tools._build_from_header`` etc.) keep resolving. Tests that
# monkeypatch the EXECUTING module (``_build_service`` / ``_emit_reauth_signal``)
# must target ``evolve_admin.google_service`` — that's where the tool logic runs.
from ..google_service import (  # noqa: F401 — re-exported for back-compat
    GMAIL_SCOPES_SEND,
    GMAIL_SCOPES_READ,
    GMAIL_SCOPES_MODIFY,
    DRIVE_SCOPES_FILE,
    DRIVE_SCOPES_READ,
    CALENDAR_SCOPES_FULL,
    CALENDAR_SCOPES_READ,
    _emit_reauth_signal,
    _load_credentials_or_signal,
    _resolve_sender_name,
    _build_from_header,
    _build_signature,
    _bot_cfg,
    _build_service,
    _header,
    _decode_gmail_body,
)


# ── Bridge tool handlers ───────────────────────────────────────────────────────
# Each resolves the bot from the bridge session/registry (the operator-surface
# ``bot`` argument), then delegates to the shared curated tool logic. _resolve_bot
# is imported lazily to avoid a circular import (tools.py imports this module at
# load time to populate TOOL_HANDLERS).


async def tool_gmail_send(arguments: dict, registry: BotRegistry) -> dict:
    """Send email from the bot's Workspace mailbox. See google_service.gmail_send."""
    from .tools import _resolve_bot
    return google_service.gmail_send(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_drive_write_file(arguments: dict, registry: BotRegistry) -> dict:
    """Create a file in the bot's Drive. See google_service.drive_write_file."""
    from .tools import _resolve_bot
    return google_service.drive_write_file(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_calendar_create_event(arguments: dict, registry: BotRegistry) -> dict:
    """Create a Calendar event. See google_service.calendar_create_event."""
    from .tools import _resolve_bot
    return google_service.calendar_create_event(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_gmail_list_messages(arguments: dict, registry: BotRegistry) -> dict:
    """List messages in the bot's mailbox. See google_service.gmail_list_messages."""
    from .tools import _resolve_bot
    return google_service.gmail_list_messages(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_gmail_get_message(arguments: dict, registry: BotRegistry) -> dict:
    """Fetch one Gmail message. See google_service.gmail_get_message."""
    from .tools import _resolve_bot
    return google_service.gmail_get_message(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_calendar_list_events(arguments: dict, registry: BotRegistry) -> dict:
    """List calendar events. See google_service.calendar_list_events."""
    from .tools import _resolve_bot
    return google_service.calendar_list_events(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_drive_list_files(arguments: dict, registry: BotRegistry) -> dict:
    """List Drive files. See google_service.drive_list_files."""
    from .tools import _resolve_bot
    return google_service.drive_list_files(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_drive_read_file(arguments: dict, registry: BotRegistry) -> dict:
    """Fetch a Drive file's content. See google_service.drive_read_file."""
    from .tools import _resolve_bot
    return google_service.drive_read_file(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_gmail_list_labels(arguments: dict, registry: BotRegistry) -> dict:
    """List Gmail labels. See google_service.gmail_list_labels."""
    from .tools import _resolve_bot
    return google_service.gmail_list_labels(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_gmail_label_message(arguments: dict, registry: BotRegistry) -> dict:
    """Add/remove labels on a message. See google_service.gmail_label_message."""
    from .tools import _resolve_bot
    return google_service.gmail_label_message(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_gmail_archive_message(arguments: dict, registry: BotRegistry) -> dict:
    """Archive a message. See google_service.gmail_archive_message."""
    from .tools import _resolve_bot
    return google_service.gmail_archive_message(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_gmail_delete_message(arguments: dict, registry: BotRegistry) -> dict:
    """Permanently delete a message. See google_service.gmail_delete_message."""
    from .tools import _resolve_bot
    return google_service.gmail_delete_message(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_gmail_trash_message(arguments: dict, registry: BotRegistry) -> dict:
    """Move a message to Trash (recoverable). See google_service.gmail_trash_message."""
    from .tools import _resolve_bot
    return google_service.gmail_trash_message(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_gmail_mark_read(arguments: dict, registry: BotRegistry) -> dict:
    """Mark a message read. See google_service.gmail_mark_read."""
    from .tools import _resolve_bot
    return google_service.gmail_mark_read(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_gmail_mark_unread(arguments: dict, registry: BotRegistry) -> dict:
    """Mark a message unread. See google_service.gmail_mark_unread."""
    from .tools import _resolve_bot
    return google_service.gmail_mark_unread(_resolve_bot(arguments.get("bot"), registry), arguments)


async def tool_drive_search(arguments: dict, registry: BotRegistry) -> dict:
    """Search the bot's full visible Drive. See google_service.drive_search."""
    from .tools import _resolve_bot
    return google_service.drive_search(_resolve_bot(arguments.get("bot"), registry), arguments)

File: packages/plugin/src/tools/GoogleTools.ts
/**
 * GoogleTools — the bot-facing curated Google tool surface (P1).
 *
 * Spec: internal/spec-google-integration-architecture-2026-06-20.md (§4.1 delivery
 * shape). Registers the 15 curated Gmail/Calendar/Drive tools on a bot whose
 * `google_integration` is configured, each proxying to the admin daemon's
 * bot-facing route:
 *
 *     bot agent  →  gmail_list_messages / drive_search / ...  (this plugin)
 *       └─ POST /api/google/call {tool, args}  over the admin-daemon UNIX SOCKET
 *            └─ admin server binds the calling bot from the socket PEER UID
 *               (never from a request field), loads creds as that bot, runs the
 *               curated tool, returns results. Creds never leave the evolve user.
 *
 * Why the unix socket (not TCP :5050 like PodStateTools): the route binds the
 * bot's identity from the kernel-reported peer uid of the socket connection.
 * That is the cross-bot-safe identity primitive — a bot literally cannot
 * present another bot's uid. The plugin therefore carries NO botId in the
 * request; the server derives it. (See web/google_bot_routes.py.)
 *
 * Eligibility is decided synchronously at register() time by
 * `googleConfiguredForBot` (reads the bot's google_integration from
 * network.json) — a bot with no Google config registers ZERO tools.
 */

import { request as httpRequest } from "http";
import { readFileSync } from "fs";
import { join } from "path";

import { Type } from "@sinclair/typebox";

import type { PluginLogger } from "openclaw/plugin-sdk/types";

const SOCKET_NAME = "admin-daemon.sock";
const ADMIN_TIMEOUT_MS = 30_000;

const SUPPORTED_MODES = new Set(["service_account_dwd", "free_gmail_oauth"]);

interface AdminResponse {
  status: number;
  body: any;
}

/**
 * POST JSON to the admin daemon over its unix socket. The socket carries the
 * caller's peer uid, which the server maps to this bot — so no identity is
 * sent in the body.
 */
function adminSocketPost(
  socketPath: string,
  urlPath: string,
  payload: unknown,
): Promise<AdminResponse> {
  return new Promise((resolve, reject) => {
    const data = Buffer.from(JSON.stringify(payload ?? {}), "utf8");
    const req = httpRequest(
      {
        socketPath,
        path: urlPath,
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Content-Length": String(data.length),
          Accept: "application/json",
        },
      },
      (res) => {
        let buf = "";
        res.on("data", (chunk: Buffer) => {
          buf += chunk.toString();
        });
        res.on("end", () => {
          let parsed: any = null;
          if (buf) {
            try {
              parsed = JSON.parse(buf);
            } catch {
              parsed = { raw: buf };
            }
          }
          resolve({ status: res.statusCode ?? 0, body: parsed });
        });
      },
    );
    req.on("error", (err: Error) => reject(err));
    req.setTimeout(ADMIN_TIMEOUT_MS, () => {
      req.destroy(new Error("admin request timed out"));
    });
    req.write(data);
    req.end();
  });
}

function textResult(text: string, isError = false) {
  return {
    content: [{ type: "text" as const, text }],
    ...(isError ? { isError: true } : {}),
  };
}

/**
 * Read this bot's google_integration mode from network.json, synchronously.
 *
 * The eligibility gate runs at register() time (which must be synchronous —
 * OC collects the toolset at session start), so we read the world-readable,
 * secret-free network.json directly rather than awaiting an HTTP probe. The
 * SERVER remains the security boundary (it re-checks config + binds identity);
 * this is only the visibility gate. Returns null on any read/parse error or
 * an unconfigured / unsupported mode — fail-closed (no tools).
 */
export function googleConfiguredForBot(
  sharedDir: string,
  botId: string,
): string | null {
  try {
    const raw = readFileSync(join(sharedDir, "network.json"), "utf8");
    const net = JSON.parse(raw);
    const gi = net?.bots?.[botId]?.google_integration;
    const mode = gi?.mode;
    return typeof mode === "string" && SUPPORTED_MODES.has(mode) ? mode : null;
  } catch {
    return null;
  }
}

// ── Tool catalog ──────────────────────────────────────────────────────────────
// Each entry mirrors a curated tool in evolve_admin/google_service.py. The
// `name` is what the bot's agent calls; the server dispatches on the same name.
// Parameter schemas are deliberately permissive (the server does the real
// validation) but describe the arguments so the agent uses them correctly.

const Str = (description: string) => Type.Optional(Type.String({ description }));
const StrReq = (description: string) => Type.String({ description });
const IntOpt = (description: string) => Type.Optional(Type.Integer({ description }));

interface GoogleToolDef {
  name: string;
  description: string;
  params: ReturnType<typeof Type.Object>;
}

const TOOL_DEFS: GoogleToolDef[] = [
  // ── Gmail read ──
  {
    name: "gmail_list_messages",
    description:
      "List messages in your Gmail mailbox. Optional `q` (Gmail search, e.g. " +
      "\"from:alice newer_than:7d\"), `label_ids`, `max_results` (default 25).",
    params: Type.Object({
      q: Str("Gmail search query, e.g. 'from:alice newer_than:7d'"),
      label_ids: Type.Optional(Type.Array(Type.String(), {
        description: "Gmail label IDs to filter on, e.g. ['INBOX','UNREAD']",
      })),
      max_results: IntOpt("Max messages to return (default 25, max 500)"),
    }),
  },
  {
    name: "gmail_get_message",
    description: "Fetch one Gmail message with headers and decoded body.",
    params: Type.Object({ id: StrReq("Gmail message id (from gmail_list_messages)") }),
  },
  {
    name: "gmail_list_labels",
    description:
      "List Gmail labels (system + user) with their ids — use this to learn " +
      "label ids before labeling/archiving.",
    params: Type.Object({}),
  },
  // ── Gmail write / modify ──
  {
    name: "gmail_send",
    description:
      "Send an email from your mailbox. Requires `to`, `subject`, `body`. " +
      "Optional `cc`, `bcc`, `attachments`.",
    params: Type.Object({
      to: Type.Union([Type.String(), Type.Array(Type.String())], {
        description: "Recipient address(es)",
      }),
      subject: StrReq("Email subject"),
      body: StrReq("Plain-text body"),
      cc: Type.Optional(Type.Union([Type.String(), Type.Array(Type.String())])),
      bcc: Type.Optional(Type.Union([Type.String(), Type.Array(Type.String())])),
      attachments: Type.Optional(Type.Array(Type.String(), {
        description:
          "File paths inside your workspace, absolute or workspace-relative " +
          "(outside paths refused); 20MB total.",
      })),
    }),
  },
  {
    name: "gmail_label_message",
    description:
      "Add and/or remove labels on a message. At least one of `add_label_ids` " +
      "/ `remove_label_ids` is required (use gmail_list_labels for ids).",
    params: Type.Object({
      id: StrReq("Gmail message id"),
      add_label_ids: Type.Optional(Type.Array(Type.String())),
      remove_label_ids: Type.Optional(Type.Array(Type.String())),
    }),
  },
  {
    name: "gmail_archive_message",
    description: "Archive a message (remove it from the inbox).",
    params: Type.Object({ id: StrReq("Gmail message id") }),
  },
  {
    name: "gmail_mark_read",
    description: "Mark a message as read.",
    params: Type.Object({ id: StrReq("Gmail message id") }),
  },
  {
    name: "gmail_mark_unread",
    description: "Mark a message as unread.",
    params: Type.Object({ id: StrReq("Gmail message id") }),
  },
  {
    name: "gmail_delete_message",
    description:
      "PERMANENTLY delete a message (not Trash — unrecoverable). Requires " +
      "`confirm: true`. For a recoverable delete, use gmail_trash_message.",
    params: Type.Object({
      id: StrReq("Gmail message id"),
      confirm: Type.Boolean({ description: "Must be true to delete" }),
    }),
  },
  {
    name: "gmail_trash_message",
    description:
      "Move a message to Trash — recoverable (~30 days, then purged). This " +
      "is the recoverable delete; gmail_label_message will not apply the " +
      "TRASH label. Requires `confirm: true`.",
    params: Type.Object({
      id: StrReq("Gmail message id"),
      confirm: Type.Boolean({ description: "Must be true to trash" }),
    }),
  },
  // ── Calendar ──
  {
    name: "calendar_list_events",
    description:
      "List upcoming calendar events. Optional `calendar_id` (default " +
      "'primary'), `time_min`/`time_max` (ISO 8601), `q`, `max_results`.",
    params: Type.Object({
      calendar_id: Str("Calendar id (default 'primary')"),
      time_min: Str("ISO 8601 lower bound (default now)"),
      time_max: Str("ISO 8601 upper bound"),
      q: Str("Free-text search"),
      max_results: IntOpt("Max events (default 25)"),
    }),
  },
  {
    name: "calendar_create_event",
    description:
      "Create a calendar event. Requires `summary`, `start`, `end` (ISO 8601 " +
      "with timezone). Optional `description`, `location`, `calendar_id`, `attendees`.",
    params: Type.Object({
      summary: StrReq("Event title"),
      start: StrReq("ISO 8601 start with timezone, e.g. 2026-06-15T18:00:00-07:00"),
      end: StrReq("ISO 8601 end with timezone"),
      description: Str("Event description"),
      location: Str("Event location"),
      calendar_id: Str("Calendar id (default 'primary')"),
      attendees: Type.Optional(Type.Array(Type.String(), {
        description: "Attendee email addresses",
      })),
    }),
  },
  // ── Drive ──
  {
    name: "drive_list_files",
    description:
      "List Drive files you own or are shared on (drive.file scope). Optional " +
      "`q` (Drive query), `page_size`, `order_by`.",
    params: Type.Object({
      q: Str("Drive query, e.g. \"mimeType='application/pdf'\""),
      page_size: IntOpt("Max files (default 25)"),
      order_by: Str("e.g. 'modifiedTime desc'"),
    }),
  },
  {
    name: "drive_read_file",
    description:
      "Fetch a Drive file's content (text decoded; binary as base64). Google-" +
      "native Docs/Sheets/Slides are not yet supported.",
    params: Type.Object({
      id: StrReq("Drive file id (from drive_list_files)"),
      max_bytes: IntOpt("Max bytes to load (default 1 MiB, max 10 MiB)"),
    }),
  },
  {
    name: "drive_write_file",
    description:
      "Create a file in your Drive. Requires `name`, `content`. Optional " +
      "`mime_type`, `parent_folder_id`.",
    params: Type.Object({
      name: StrReq("File name"),
      content: StrReq("File body (text)"),
      mime_type: Str("MIME type (default text/plain)"),
      parent_folder_id: Str("Parent Drive folder id"),
    }),
  },
  {
    name: "drive_search",
    description:
      "Search across ALL Drive files visible to you (drive.readonly scope) — " +
      "broader than drive_list_files. Optional `q`, `page_size`, `order_by`.",
    params: Type.Object({
      q: Str("Drive query, e.g. \"name contains 'invoice'\""),
      page_size: IntOpt("Max files (default 25)"),
      order_by: Str("e.g. 'modifiedTime desc'"),
    }),
  },
];

/**
 * Build the curated Google tool factories for a configured bot.
 *
 * Each tool POSTs {tool, args} to /api/google/call over the unix socket. The
 * server binds identity from the peer uid — the plugin sends NO botId. A 403
 * `not_configured` (config changed out from under us) surfaces as a clean
 * tool error rather than a crash.
 */
export function createGoogleToolFactories(
  config: { sharedDir: string; botId: string },
  logger: PluginLogger,
): Array<(ctx: Record<string, unknown>) => unknown> {
  const socketPath = join(config.sharedDir, SOCKET_NAME);

  return TOOL_DEFS.map((def) => (_ctx: Record<string, unknown>) => ({
    name: def.name,
    description: def.description,
    parameters: def.params,
    async execute(_toolCallId: string, rawParams: unknown) {
      const args = (rawParams ?? {}) as Record<string, unknown>;
      try {
        const res = await adminSocketPost(socketPath, "/api/google/call", {
          tool: def.name,
          args,
        });
        if (res.status === 200) {
          return textResult(JSON.stringify(res.body));
        }
        const err =
          (res.body && (res.body.detail || res.body.error)) ||
          `HTTP ${res.status}`;
        return textResult(`${def.name} failed: ${err}`, true);
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        logger.warn(`${def.name}: admin request failed: ${msg}`);
        return textResult(`${def.name}: ${msg}`, true);
      }
    },
  }));
}
